components/core/src/clp/BufferedFileReader.cpp

```diff
@@ -22,133 +29,59 @@
  * @return ErrorCode_EndOfFile on EOF
  * @return ErrorCode_Success on success
  */
-auto read_into_buffer(int fd, char* buf, size_t num_bytes_to_read, size_t& num_bytes_read)
-        -> ErrorCode;
-
-auto read_into_buffer(int fd, char* buf, size_t num_bytes_to_read, size_t& num_bytes_read)
-        -> ErrorCode {
+auto read_into_buffer(
+        ReaderInterface* reader,
+        char* buf,
+        size_t num_bytes_to_read,
+        size_t& num_bytes_read
+) -> ErrorCode;
+
+auto read_into_buffer(
+        ReaderInterface* reader,
+        char* buf,
+        size_t num_bytes_to_read,
+        size_t& num_bytes_read
+) -> ErrorCode {
     num_bytes_read = 0;
-    while (true) {
-        auto const bytes_read = ::read(fd, buf, num_bytes_to_read);
-        if (0 == bytes_read) {
-            break;
-        }
-        if (bytes_read < 0) {
-            return ErrorCode_errno;
-        }
-
-        buf += bytes_read;
-        num_bytes_read += bytes_read;
-        num_bytes_to_read -= bytes_read;
-        if (num_bytes_read == num_bytes_to_read) {
-            return ErrorCode_Success;
-        }
-    }
+    auto const error_code = reader->try_read(buf, num_bytes_to_read, num_bytes_read);
     if (0 == num_bytes_read) {
         return ErrorCode_EndOfFile;
     }
-    return ErrorCode_Success;
+    return error_code;
 }
 }  // namespace
 
-BufferedFileReader::BufferedFileReader(size_t base_buffer_size) {
+BufferedFileReader::BufferedFileReader(
+        std::unique_ptr<ReaderInterface> reader_interface,
+        size_t base_buffer_size
+)
+        : m_file_reader(std::move(reader_interface)) {
     if (base_buffer_size % cMinBufferSize != 0) {
         throw OperationFailed(ErrorCode_BadParam, __FILENAME__, __LINE__);
     }
     m_base_buffer_size = base_buffer_size;
     m_buffer.resize(m_base_buffer_size);
-}
-
-BufferedFileReader::~BufferedFileReader() {
-    close();
-}
 
-auto BufferedFileReader::try_open(string const& path) -> ErrorCode {
-    // Cleanup in case caller forgot to call close before calling this function
-    close();
-
-    m_fd = ::open(path.c_str(), O_RDONLY);
-    if (-1 == m_fd) {
-        if (ENOENT == errno) {
-            return ErrorCode_FileNotFound;
-        }
-        return ErrorCode_errno;
-    }
-    m_path = path;
-    m_file_pos = 0;
+    m_file_pos = m_file_reader->get_pos();
     m_buffer_begin_pos = 0;
-    m_buffer_reader.emplace(m_buffer.data(), 0);
+    m_buffer_reader = BufferReader{m_buffer.data(), 0};
     m_highest_read_pos = 0;
```

Comment:
This is not really necessary given how the code works. But it might be more clear (or less confusing) to set it to m_file_pos so I will take this change